### PR TITLE
NOMMU: Add option to exec ELF from different map than file's

### PR DESCRIPTION
**Pull request by agkaminski**
JIRA: RTOS-267

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [x] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (stm32l4).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work:
https://github.com/phoenix-rtos/phoenix-rtos-project/pull/536
https://github.com/phoenix-rtos/phoenix-rtos-utils/pull/158
https://github.com/phoenix-rtos/plo/pull/215
https://github.com/phoenix-rtos/libphoenix/pull/210
https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/348

### CODE REVIEW

File: proc/process.c [hunk 1]
```diff
@@ -578,13 +585,26 @@ int process_load(process_t *process, vm_object_t *o, offs_t base, size_t size, v
...
 		if (phdr->p_flags & PF_R)
 			prot |= PROT_READ;
 
-		if (phdr->p_flags & PF_X)
+		if (phdr->p_flags & PF_X) {
 			prot |= PROT_EXEC;
 
+			if (process->imapp != NULL &&
+					(((ptr_t)base < (ptr_t)process->imapp->start) ||
+					((ptr_t)base > (ptr_t)process->imapp->stop))) {
```
Comment (HBuczynski) [inline proc/process.c:598]: You don't check whether the size of the elf fits in the map boundaries.

File: proc/process.c [hunk 2]
```diff
@@ -94,8 +95,13 @@ static void process_destroy(process_t *p)
 
 	posix_died(p->id, p->exit);
 
-	if (p->mapp != NULL)
+	if (p->mapp != NULL) {
 		vm_mapDestroy(p, p->mapp);
+	}
+
+	if (p->imapp != NULL) {
+		vm_mapDestroy(p, p->imapp);
```
Comment (HBuczynski) [inline proc/process.c:103]: What do you think about destroying the `imapp` here, after `fork`:
https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/8b5a890117bd49e34c8c1cc750fdf7685ad23daa/proc/process.c#L1225

File: proc/process.c (hunk 1 above)
Comment (agkaminski) [reply proc/process.c:598]: I don't really have to, as `vm_mmap` below will verify that for me

File: proc/process.c [hunk 3]
```diff
@@ -288,6 +294,7 @@ int proc_start(void (*initthr)(void *), void *arg, const char *path)
 #endif
 
 	proc_changeMap(process, NULL, NULL);
+	process->imapp = NULL;
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply proc/process.c:297]: No need in `proc_vforkedExit`, it's just a pointer and we're ending the process, fixed in process_execve

File: proc/process.c (hunk 2 above)
Comment (agkaminski) [reply proc/process.c:103]: Yes, thanks for spotting that

**Review by HBuczynski — APPROVED** [review]
LGTM

File: hal/armv7a/cpu.c [hunk 4]
```diff
@@ -199,3 +199,14 @@ char *hal_cpuFeatures(char *features, unsigned int len)
...
+
+/* cache management */
+
+
+void hal_cleanDCache(ptr_t start, size_t len)
+{
+	(void)start;
+	(void)len;
+	/* TODO */
+}
```
Comment (HBuczynski) [inline hal/armv7a/cpu.c:212]: Cortex-A has already defined function for data cache cleaning:
https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/ae0a22325e93d3f9d921a1feffd089e52de1463b/hal/armv7a/armv7a.h#L53-L54

File: hal/armv7m/cpu.c [hunk 5]
```diff
@@ -255,6 +255,23 @@ void hal_wdgReload(void)
...
+{
+	(void)start;
+	(void)len;
+
+#ifdef CPU_IMXRT
+	/* Clean the whole cache, no harm in being overzealous */
+	_imxrt_cleanDCache();
+#else
+	/* TODO */
+#endif
```
Comment (HBuczynski) [inline hal/armv7m/cpu.c:271]: Data cache cleaning for Cortex-M is the same for `imxrt` and `stm32`. I think that the `ifdef` is not needed here and the code can be common for all Cortex-M7 architectures.

Comment (agkaminski) [reply hal/armv7m/cpu.c:271]: Can't really be done due to code arch - scb address is kept in x_common structs per arch

File: hal/armv7a/cpu.c (hunk 4 above)
Comment (agkaminski) [reply hal/armv7a/cpu.c]: Done

File: hal/armv7m/cpu.c (hunk 5 above)
Comment (HBuczynski) [reply hal/armv7m/cpu.c:271]: If you look at functions using `scb` registers in `imxrt106x`, `imxrt117x` and `stm32`, they are the same. There is copy-paste code between 3 targets.

**Review by MaciejPurski — APPROVED** [review]
HW tested on imxrt106x.

File: hal/armv7m/cpu.c (hunk 5 above)
Comment (HBuczynski) [reply hal/armv7m/cpu.c:271]: Yeah, I wanted to raise a topic to avoid copy-pasting in the future. Maybe put a `FIXME` comment to not forget about it.

**Comment by agkaminski** [comment]
HW test fail due to change in spawnSyspage syscall API

**Pull request merged**